yew-functional/src/lib.rs

```diff
@@ -381,45 +249,31 @@ where
 
         // Initialize hook if this is the first call
         if hook_pos >= hook_state.hooks.len() {
-            let initial_state = Rc::new(RefCell::new(initial_state_producer()));
+            let initial_state = Rc::new(RefCell::new(initializer()));
             hook_state.hooks.push(initial_state.clone());
             hook_state.destroy_listeners.push(Box::new(move || {
                 initial_state.borrow_mut().deref_mut().tear_down();
             }));
         }
 
-        let hook = hook_state.hooks[hook_pos].clone();
+        let hook = hook_state
+            .hooks
+            .get(hook_pos)
+            .expect("Not the same number of hooks. Hooks must not be called conditionally")
+            .clone();
 
-        (hook, hook_state.process_message.clone())
+        HookUpdater {
+            hook,
+            process_message: hook_state.process_message.clone(),
+        }
     });
 
-    let hook_callback = {
-        let hook = hook.clone();
-        Box::new(move |update: HookUpdate, post_render| {
-            let hook = hook.clone();
-            process_message(
-                Box::new(move || {
-                    let mut hook = hook.borrow_mut();
-                    let hook = hook.downcast_mut::<InternalHookState>();
-                    let hook = hook.expect(
-                        "Incompatible hook type. Hooks must always be called in the same order",
-                    );
-                    update(hook)
-                }),
-                post_render,
-            );
-        })
-    };
-    let mut hook = hook.borrow_mut();
-    let hook = hook.downcast_mut::<InternalHookState>();
-    let mut hook =
-        hook.expect("Incompatible hook type. Hooks must always be called in the same order");
-
     // Execute the actual hook closure we were given. Let it mutate the hook state and let
     // it create a callback that takes the mutable hook state.
-    hook_runner(&mut hook, hook_callback)
-}
+    let mut hook = updater.hook.borrow_mut();
+    let hook: &mut InternalHook = hook
+        .downcast_mut()
+        .expect("Incompatible hook type. Hooks must always be called in the same order");
```

Comment:
nit: Error messages should be lowercase without trailing punctuation. This applies to a lot of places

I realize a lot of these messages were kept as-is but we might as well fix it now.

Reply:
Did my best, feel free to take a look.